### PR TITLE
feat(ingest): implement conditional updates for rest endpoints

### PR DESCRIPTION
**Pull request by djordje-mijatovic**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-io/src/main/java/com/linkedin/metadata/params/ExtraIngestParamsHelper.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+@Slf4j
+public class ExtraIngestParamsHelper {
+
+    // condUpdate String format: "urn1+AspectName1=createdOn1;urn2+AspectName2=createdOn2;...;urnN+AspectNameN=createdOnN"
+
+    public static Map<String, Long> extractCondUpdate(String condUpdate) {
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/params/ExtraIngestParamsHelper.java:14]: As discussed, we should add a javadoc to explain the format of the `createdOn` parameters.

**Review by mattmatravers — COMMENTED** [review]
This is a continuation of https://github.com/datahub-project/datahub/pull/6652 where Djordje's branch got into a strange state, so we've started afresh with a new branch.

File: metadata-io/src/main/java/com/linkedin/metadata/params/ExtraIngestParams.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+@Builder
+public class ExtraIngestParams {
+
+    private Map<String, Long> createdOnMap;
+
+    public Long getCreatedOn(String urn, String aspectName) {
+        if (urn == null || aspectName == null || createdOnMap == null) {
+            return null;
+        }
+        return createdOnMap.containsKey(urn + "+" + aspectName) ? createdOnMap.get(urn + "+" + aspectName) : null;
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/params/ExtraIngestParams.java:18]: Does this simplify?

```suggestion
        return createdOnMap.get(urn + "+" + aspectName);
```

**Review by mattmatravers — APPROVED** [review]
Looks good from my perspective. Ready for Acryl to review.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 3]
```diff
@@ -1697,7 +1705,7 @@ public RollbackResult deleteAspect(String urn, String aspectName, @Nonnull Map<S
         latest.setCreatedOn(survivingAspect.getCreatedOn());
         latest.setCreatedBy(survivingAspect.getCreatedBy());
         latest.setCreatedFor(survivingAspect.getCreatedFor());
-        _aspectDao.saveAspect(latest, false);
+        _aspectDao.saveAspect(latest, false, null);
```
Comment (RyanHolstien) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:1708]: Confused on why some of these are passing null through even though the method signature was updated to have the extraIngestParams

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 4]
```diff
@@ -748,7 +753,7 @@ public RecordTemplate ingestAspectIfNotPresent(@Nonnull Urn urn, @Nonnull String
         long nextVersion = _aspectDao.getNextVersion(urnStr, aspectName);
 
         return ingestAspectToLocalDBNoTransaction(urn, aspectName, ignored -> newValue, auditStamp,
-            internalSystemMetadata, latest, nextVersion);
+            internalSystemMetadata, latest, nextVersion, null);
```
Comment (RyanHolstien) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:756]: Are we assuming some methods of ingestion are not going to pass these down ever? Does it not make sense for this flow?

Comment (djordje-mijatovic) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:756]: @RyanHolstien We did not implemente conditional updated for all flows, but just for few endpoints. For example, conditional update is not supported for GraphQL nor delete operation for example. Those two cases you have mentioned are not part of the flows we have choose to start from.

**Comment by jjoyce0510** [comment]
Hi @djordje-mijatovic--

Are you still active on this PR? 

John

**Comment by jjoyce0510** [comment]
Closing for now, will reopen once we decide to proceed with conditional updates.

**Pull request closed without merging**